R/replace.R

```diff
@@ -180,18 +182,50 @@ str_replace_na <- function(string, replacement = "NA") {
 
 str_transform <- function(string, pattern, replacement) {
   loc <- str_locate(string, pattern)
-  str_sub(string, loc, omit_na = TRUE) <- replacement(str_sub(string, loc))
+  new <- replacement(str_sub(string, loc))
+  str_sub(string, loc, omit_na = TRUE) <- new
   string
 }
-str_transform_all <- function(string, pattern, replacement) {
+
+str_transform_all <- function(string, pattern, replacement, error_call = caller_env()) {
   locs <- str_locate_all(string, pattern)
 
-  for (i in seq_along(string)) {
-    for (j in rev(seq_len(nrow(locs[[i]])))) {
-      loc <- locs[[i]]
-      str_sub(string[[i]], loc[j, 1], loc[j, 2]) <- replacement(str_sub(string[[i]], loc[j, 1], loc[j, 2]))
-    }
+  old <- str_sub_all(string, locs)
+
+  # unchop list into a vector, apply replacement(), and then rechop back into
+  # a list
+  old_flat <- vctrs::list_unchop(old)
+  if (length(old_flat) == 0) {
+    # minor optimisation to avoid problems with the many replacement
+    # functions that use paste
+    new_flat <- character()
+  } else {
+    new_flat <- replacement(old_flat)
```

Comment:
@DavisVaughan do you think it's worth wrapping this to provide a more informative error? (something like `replacement` function must be vectorised?)

Reply:
Are you talking about removing the `length(old_flat) == 0` special case handling?

That was what I mentioned back here, I think 
https://github.com/tidyverse/stringr/pull/462#discussion_r987127244

We went through a similar thing with `dplyr::rename_with()` where we made it so that `.fn` has to return a vector of length equal to the input length (i.e. typical size stability, like here).

For dplyr, we did get a few questions about this, so we added a `paste0(recycle0 = TRUE)` example:
https://github.com/tidyverse/dplyr/pull/6690

I think you could do a few things to mitigate issues:
- Add an example of `paste0`, like dplyr
- You already have a error below when `length(new_flat) != length(old_flat)`, but you could specially call out that `replacement` needs to be _vectorized_ there as well.
- If you're feeling particularly friendly, you could check to see if `replacement` was identical to `paste` or `paste0` in the check below after `length(new_flat) != length(old_flat)` fails. If so, you could throw a special error message just for that very common case, and probably advise to use `recycle0`.

Reply:
If you look at the revdeps, a lot of the errors are folks passing vectors to functions that expect scalars. So I was thinking it might make sense to wrap the call to `replacement()` to provide some hint in the error message.

Reply:
Ah like

```r
    Error in if (x > 0) "+" else "-" : the condition has length > 1
    Calls: <Anonymous> ... <Anonymous> -> str_transform_all -> replacement -> <Anonymous>
```

Ok now I understand what you mean.

I think it would be quite hard to make that error precise, i.e. you can't tell when they've provided a function that only expects a scalar very easily.

So the alternative is to _always_ wrap any error from `replacement()` with some general advice that talks needing a vectorized function. I think that could be useful if you frame it like:

```
Error in str_fn():
Failed to apply `replacement()`. Check the following in addition to the error thrown by `replacement()`:
- `replacement()` must be a vectorized function capable of handling input of any length.
- `replacement()` must return a character vector with output the same length as the input.
Caused by error in replacement():
<parent error here>
```

Reply:
Yeah, exactly!